### PR TITLE
executor: change sort to parallel

### PR DESCRIPTION
**Pull request by yu34po**
### What is changed and how it works?
make sort executor to parallel process
how it works:
1.get all rows from children executor
2.start several sort worker goroutine, average split rows to every sort worker, for every sort worker sort rows of itself
3.merge the sorted rows from every sort worker
 
### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `tidb-ansible` repository
 - Need to be included in the release note

### CODE REVIEW

File: executor/merge_sort.go [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+	schema  *expression.Schema
+
+	keyExprs []expression.Expression
+	keyTypes []*types.FieldType
+	// keyColumns is the column index of the by items.
+	keyColumns []int
+	// keyCmpFuncs is used to compare each ByItem.
+	keyCmpFuncs []chunk.CompareFunc
+	// keyChunks is used to store ByItems values when not all ByItems are column.
+	keyChunks *chunk.List
```
Comment (XuHuaiyu) [inline executor/merge_sort.go]: It seems that we do not need this after https://github.com/pingcap/tidb/pull/9197.

File: sessionctx/variable/session.go [hunk 2]
```diff
@@ -759,6 +760,9 @@ type Concurrency struct {
 
 	// IndexSerialScanConcurrency is the number of concurrent index serial scan worker.
 	IndexSerialScanConcurrency int
+
+	//MergeSort is the number of concurrent sort worker
+	MergeSortConcurrency int
```
Comment (XuHuaiyu) [inline sessionctx/variable/session.go]: Should we add a definition for this in `sessionctx/variable/sysvar.go`

File: executor/merge_sort.go [hunk 3]
```diff
@@ -0,0 +1,341 @@
+// Copyright 2017 PingCAP, Inc.
```
Comment (winoros) [inline executor/merge_sort.go]: 2019

File: executor/merge_sort.go [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"github.com/pingcap/tidb/util"
```
Comment (winoros) [inline executor/merge_sort.go]: move down to the single import group

File: executor/merge_sort.go [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+	workerRowLen []int
+	workerRowIdx []int
+
+	memTracker    *memory.Tracker
+	concurrency   int
+	allColumnExpr bool
+}
+
+// SortWorker represents worker routine to process sort
+type SortWorker struct {
```
Comment (winoros) [inline executor/merge_sort.go]: No need to be exported?

File: executor/merge_sort.go [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+
+	workerRowLen []int
+	workerRowIdx []int
+
+	memTracker    *memory.Tracker
+	concurrency   int
+	allColumnExpr bool
+}
+
+// SortWorker represents worker routine to process sort
```
Comment (winoros) [inline executor/merge_sort.go]: Add `.`

File: executor/merge_sort.go [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+		sort.Slice(sw.rowPtrs, sw.keyChunksLess)
+	}
+	return
+}
+
+// Close implements the Executor Close interface.
+func (e *MergeSortExec) Close() error {
+	e.memTracker.Detach()
+	e.memTracker = nil
+	return errors.Trace(e.children[0].Close())
```
Comment (winoros) [inline executor/merge_sort.go]: We don't need to call `errors.Trace`. Just `return e.children[0].Close()` is okay now.

File: executor/merge_sort.go [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+
+		workerRowsCount := e.rowChunks.Len() / e.concurrency
+		workerIdx := e.sortWorkerIndex(workerRowsCount)
+
+		wg := &sync.WaitGroup{}
+		wg.Add(int(e.concurrency))
+		go e.wait4WorkerSort(wg, finishedCh)
+
+		for i := 0; i < e.concurrency; i++ {
+			// last worker must complete the rest of rows
```
Comment (winoros) [inline executor/merge_sort.go]: `Last` and add `.`

File: sessionctx/variable/session.go (hunk 2 above)
Comment (winoros) [inline sessionctx/variable/session.go]: Don't forget to add unit-test for this session variable.

File: executor/merge_sort.go [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+				defer wg.Done()
+				sw.run()
+			}, nil)
+		}
+
+		<-finishedCh
+		e.fetched = true
+	}
+
+	for req.NumRows() < e.maxChunkSize {
```
Comment (qw4990) [inline executor/merge_sort.go]: This PR can be rebased to the newest master branch and then this condition can be replaced by `!req.IsFull()`, which can support chunk size control(#9166)

File: executor/merge_sort.go [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+			}
+		}
+		finishedCh := make(chan struct{})
+
+		workerRowsCount := e.rowChunks.Len() / e.concurrency
+		workerIdx := e.sortWorkerIndex(workerRowsCount)
+
+		wg := &sync.WaitGroup{}
+		wg.Add(int(e.concurrency))
+		go e.wait4WorkerSort(wg, finishedCh)
```
Comment (qw4990) [inline executor/merge_sort.go]: This goroutine is unnecessary?
Just replace `<-finishedCh` to `wg.Wait()` at line 205.

File: executor/merge_sort.go [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+	for	!req.IsFull() {
+		j := 0
+		for j < e.concurrency && e.workerRowIdx[j] >= e.workerRowLen[j] {
+			j++
+		}
+		if j >= e.concurrency {
+			break
+		}
+		minRowPtr := (*e.workerRowPtrs[j])[e.workerRowIdx[j]]
+		for i := j + 1; i < e.concurrency; i++ {
```
Comment (qw4990) [inline executor/merge_sort.go]: This merge process may be optimized by using a `heap`.
The `heap` maintains `e.concurrency` values representing each worker's minimum value.
In each loop, we pop a minimum value from the `heap` as result and push a new value from the same `worker` with the popped one into the `heap`.

It's maybe a little complicated, so I prefer to do it in another PR and leave a `TODO` here.

**Comment by zz-jason** [comment]
@yu34po Sorry for the late review, please update to the master branch and resolve conflicts.

**Comment by zz-jason** [comment]
closed due to outdated, fell free to reopen it if it's needed.
